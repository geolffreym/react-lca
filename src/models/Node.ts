export const deserializeBT = (arr: (number | null)[]): Node | null => {
  if (!arr.length || arr[0] == null) {
    return null;
  }

  const stack: Node[] = [];
  const first = arr.shift()! // non-null value
  const root = new Node(first);
  stack.push(root); // Add first node as root in stack

  while (stack.length && arr.length) {
    const node = stack.shift();
    if (!node) {
      return root;
    }

    const lVal = arr.shift();
    const rVal = arr.shift();

    node.left = lVal ? new Node(lVal) : null;
    node.right = rVal ? new Node(rVal) : null;

    // Circular node memory stack reference
    // Every new node after being added to btree position (left, right) is added to the stack as a reference.
    // The memory keep a track of each node and create a "pseudo-recursive" tree generation.
    node.left && stack.push(node.left);
    node.right && stack.push(node.right);
  }

  return root;
};


/**
 * Find LCA for p and q
 * 
 * Pre-order behavior: Executing first the function over the node and then recursively left and right node.
 * @param node: Root tree node
 * @param p: Node A to find common ancestor
 * @param q: Node B to find common ancestor 
 * @returns Node if found else null
 */
export const LCA = (node: Node | null, p: number, q: number): Node | null => {
  if (!node) return null
  if (node.value === p || node.value === q) return node
  let left: Node | null = LCA(node.left, p, q)
  let right: Node | null = LCA(node.right, p, q)

  if (left && right) return node
  if (!left && !right) return null
  return left ? left : right
}

// shadowing name for typescript Node
export default class Node {
  public value: number;
  public left: Node | null = null;
  public right: Node | null = null;

  constructor(val: number) {
    this.value = val;
  }
}
