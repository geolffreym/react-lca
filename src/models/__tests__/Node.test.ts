import Node, { deserializeBT, LCA } from '../Node';

describe('deserializes BT correctly', () => {
  const dfs = (node: Node | null, preOrder: number[]): void => {
    if (!node) {
      return;
    }

    preOrder.push(node.value);
    dfs(node.left, preOrder);
    dfs(node.right, preOrder);
  };

  it('should correctly work with a raw array', () => {
    const serializedArray = [3, 9, 7, 2, 6, null, 4];
    const tree = deserializeBT(serializedArray);

    expect(tree).toMatchSnapshot();

    const preOrder: number[] = [];
    dfs(tree, preOrder);

    expect(preOrder).toStrictEqual([3, 9, 2, 6, 7, 4]);
  });

  it('should work with json array', () => {
    const stringArray = '[3, 9, 7, 2, 6, null, 4]';
    const tree = deserializeBT(JSON.parse(stringArray));

    expect(tree).toMatchSnapshot();

    const preOrder: number[] = [];
    dfs(tree, preOrder);

    expect(preOrder).toStrictEqual([3, 9, 2, 6, 7, 4]);
  });

  it('should find the LCA based on root, p,q', () => {
    const array = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22];
    const tree = deserializeBT(array);

    expect(LCA(tree, 18, 19)?.value).toStrictEqual(9)
    expect(LCA(tree, 10, 15)?.value).toStrictEqual(1)
  })

  it('should return null for invalid LCA based on root, p,q', () => {
    const array = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22];
    const tree = deserializeBT(array);

    expect(LCA(tree, 100,50)).toStrictEqual(null)
  })

});
