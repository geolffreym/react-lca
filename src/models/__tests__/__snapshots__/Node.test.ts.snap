// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deserializes BT correctly should correctly work with a raw array 1`] = `
Node {
  "left": Node {
    "left": Node {
      "left": null,
      "right": null,
      "value": 2,
    },
    "right": Node {
      "left": null,
      "right": null,
      "value": 6,
    },
    "value": 9,
  },
  "right": Node {
    "left": null,
    "right": Node {
      "left": null,
      "right": null,
      "value": 4,
    },
    "value": 7,
  },
  "value": 3,
}
`;

exports[`deserializes BT correctly should work with json array 1`] = `
Node {
  "left": Node {
    "left": Node {
      "left": null,
      "right": null,
      "value": 2,
    },
    "right": Node {
      "left": null,
      "right": null,
      "value": 6,
    },
    "value": 9,
  },
  "right": Node {
    "left": null,
    "right": Node {
      "left": null,
      "right": null,
      "value": 4,
    },
    "value": 7,
  },
  "value": 3,
}
`;
